grype/db/v6/vulnerability_decorator_store.go

```diff
@@ -12,28 +12,139 @@ import (
 )
 
 type VulnerabilityDecoratorStoreWriter interface {
-	AddKnownExploitedVulnerabilities(kevs ...*KnownExploitedVulnerabilityHandle) error
+	AddKnownExploitedVulnerabilities(...*KnownExploitedVulnerabilityHandle) error
+	AddEpss(...*EpssHandle) error
 }
 
 type VulnerabilityDecoratorStoreReader interface {
 	GetKnownExploitedVulnerabilities(cve string) ([]KnownExploitedVulnerabilityHandle, error)
+	GetEpss(cve string) ([]EpssHandle, error)
 }
 
 type vulnerabilityDecoratorStore struct {
-	db         *gorm.DB
-	blobStore  *blobStore
-	kevEnabled bool
+	db          *gorm.DB
+	blobStore   *blobStore
+	kevEnabled  bool
+	epssEnabled bool
+	epssDate    *time.Time
 }
 
 func newVulnerabilityDecoratorStore(db *gorm.DB, bs *blobStore, dbVersion schemaver.SchemaVer) *vulnerabilityDecoratorStore {
 	minSupportedKEVClientVersion := schemaver.New(6, 0, 1)
+	minSupportedEPSSClientVersion := schemaver.New(6, 0, 2)
 	return &vulnerabilityDecoratorStore{
-		db:         db,
-		blobStore:  bs,
-		kevEnabled: dbVersion.GreaterOrEqualTo(minSupportedKEVClientVersion),
+		db:          db,
+		blobStore:   bs,
+		kevEnabled:  dbVersion.GreaterOrEqualTo(minSupportedKEVClientVersion),
+		epssEnabled: dbVersion.GreaterOrEqualTo(minSupportedEPSSClientVersion),
 	}
 }
 
+func (s *vulnerabilityDecoratorStore) AddEpss(epss ...*EpssHandle) error {
+	if !s.epssEnabled {
+		// when populating a new DB we should never run into capability issues
+		return ErrDBCapabilityNotSupported
+	}
+
+	for i := range epss {
+		e := epss[i]
+
+		if err := s.db.Create(e).Error; err != nil {
+			return fmt.Errorf("unable to create EPSS: %w", err)
+		}
+
+		if err := s.setEPSSMetadata(e.Date); err != nil {
+			return fmt.Errorf("unable to set EPSS metadata: %w", err)
+		}
+	}
+	return nil
+}
+
+func (s *vulnerabilityDecoratorStore) setEPSSMetadata(date time.Time) error {
```

Comment:
Right now, this is non-deterministic. If there are multiple EPSS dates, the one processed last wins. I think from the log.Warn below that this is an unexpected situation? Is it worth making this deterministic anyway, eg only updating the existing date if the new date is newer?